src/elf/symver.rs

```diff
@@ -0,0 +1,924 @@
+            .ok()
+            .map(|ElfVersym { vs_val }| Versym { vs_val })
+    }
+}
+
+impl<'a> IntoIterator for &'_ VersymSection<'a> {
+    type Item = <VersymIter<'a> as Iterator>::Item;
+    type IntoIter = VersymIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        VersymIter {
+            bytes: self.bytes,
+            offset: 0,
+            ctx: self.ctx,
+        }
+    }
+}
+
+/// Iterator over the [`Versym`] entries from the [`SHT_GNU_VERSYM`] section.
+pub struct VersymIter<'a> {
+    bytes: &'a [u8],
+    offset: usize,
+    ctx: container::Ctx,
+}
+
+impl<'a> Iterator for VersymIter<'a> {
+    type Item = Versym;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.bytes.len() {
+            None
+        } else {
+            self.bytes
+                .gread_with::<ElfVersym>(&mut self.offset, self.ctx.le)
+                .ok()
+                .map(|ElfVersym { vs_val }| Versym { vs_val })
+                .or_else(|| {
+                    // self.bytes are not a multiple of ElfVersym.
+                    // Adjust offset to continue yielding Nones.
+                    self.offset = self.bytes.len();
+                    None
+                })
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = (self.bytes.len() - self.offset) / core::mem::size_of::<Self::Item>();
+        (len, Some(len))
+    }
+}
+
+impl ExactSizeIterator for VersymIter<'_> {}
+
+impl FusedIterator for VersymIter<'_> {}
+
+/// An ELF [Symbol Version][lsb-versym] entry.
+///
+/// [lsb-versym]: https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/symversion.html#SYMVERTBL
+#[derive(Debug)]
+pub struct Versym {
+    pub vs_val: u16,
+}
+
+impl Versym {
+    /// Returns true if the symbol is local and not available outside the object according to
+    /// [`VER_NDX_LOCAL`].
+    #[inline]
+    pub fn is_local(&self) -> bool {
+        self.vs_val == VER_NDX_LOCAL
+    }
+
+    /// Returns true if the symbol is defined in this object and globally available according
+    /// to [`VER_NDX_GLOBAL`].
+    #[inline]
+    pub fn is_global(&self) -> bool {
+        self.vs_val == VER_NDX_GLOBAL
+    }
+
+    /// Returns true if the `hidden` bit is set according to the [`VERSYM_HIDDEN`] bitmask.
+    #[inline]
+    pub fn is_hidden(&self) -> bool {
+        (self.vs_val & VERSYM_HIDDEN) == VERSYM_HIDDEN
+    }
+
+    /// Returns the symbol version index according to the [`VERSYM_VERSION`] bitmask.
+    #[inline]
+    pub fn version(&self) -> u16 {
+        self.vs_val & VERSYM_VERSION
+    }
+}
+
+/************************
+ *  Version Definition  *
+ ************************/
+
+/// Helper struct to iterate over [Version Definition][Verdef] and [Version Definition
+/// Auxiliary][Verdaux] entries.
+#[derive(Debug)]
+pub struct VerdefSection<'a> {
+    /// String table used to resolve version strings.
+    pub verstr: Strtab<'a>,
```

Comment:
so this is a `pub` struct, and this field is `pub`; I don't see it used anywhere, so I presume it's for clients to resolve symbols; but I'm wondering if we can expose an api to resolve it for them? Since once this is a `pub` it's `pub` forever basically, so just making sure it's something want to commit to.

Reply:
Right, these `Strtab`s are used to resolve any symbol version related strings.
Only the rustdoc examples at the top of the module show the usage of those.

> but I'm wondering if we can expose an api to resolve it for them

Out of my head I see the following two approaches:
1. Pass a reference to the corresponding `Strtab` down to `Verneed | Vernaux | Vernaux`, and provide APIs on those structs to get the strings.
2. Offer APIs on the sections to resolve the strings, eg `VerneedSection::file_name(_:&Verneed) -> Option<&str>`

Not really decided which one I prefer, maybe after a night of sleep :)

Reply:
This strtab must be the same as `Elf::dynstrtab` (because the dynamic loader can't use section headers to find the strings). So it might be better to pass a reference to `Elf::dynstrtab` when needed instead of having this field.

Reply:
@philipc good point, I was mainly following the LSB here.
> The sh_link member of the section header shall point to the section that contains the strings referenced by this section.

Out of interest I also look at the common static linkers, how they craft the version related sections ([gold](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=gold/layout.cc;hb=250d07de5cf6efc81ed934c25292beb63c7e3129#l4976), [ld](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=bfd/elf.c;hb=64dbf74d428e97de737c13f42767dab88493f187#l3979) and [lld](https://github.com/llvm/llvm-project/blob/b06426da764a8d0254521b33d667db8f26ae5e2f/lld/ELF/SyntheticSections.cpp#L3100)).

In that sense we could drop the `verstr` fields completely for now and only offer the minimal APIs. Users could then  resolve the strings against `Elf::dynstrtab` directly.

As discussed previously, we could then look into providing higher-level APIs in a second PR later (which I started to toy around with).

Reply:
Committed that change for now and updated the doc examples.
Let's see your opinions.